File: JavaScript/0371. Sum of Two Integers.js
// Calculate the sum of two integers a and b, but you are not allowed to use the operator + and -.
//
// Example:
// Given a = 1 and b = 2, return 3.

/**
 * @param {number} a
 * @param {number} b
 * @return {number}
 */
/** 1) Iteration */
// e.g. 3 + 1
// 0b11 & 0b01 = 0b01
// 0b11 ^ 0b01 = 0b10
function getSum1(a, b) {
  while (b !== 0) {
    const carry = a & b; // carry over
    a = a ^ b; // sum without carry over
    b = carry << 1; // move the carry over to the correct position
  }

  return a;
}

/** 2) Recursion */
function getSum(a, b) {
  if (b === 0) return a;
  return getSum(a ^ b, (a & b) << 1);
}

File: Python/0371. Sum of Two Integers.py
# Given two integers a and b, return the sum of the two integers without using the operators + and -.
#
# Example 1:
#
# Input: a = 1, b = 2
# Output: 3
#
# Example 2:
#
# Input: a = 2, b = 3
# Output: 5
#
# Constraints:
#
# -1000 <= a, b <= 1000


# 1)
class Solution:
    def getSum(self, a: int, b: int) -> int:
        l = [a, b]
        return sum(l)


# 2) Bit
# https://leetcode.com/problems/sum-of-two-integers/discuss/776952/Python-BEST-LeetCode-371-Explanation-for-Python
